src/olympia/amo/celery.py

```diff
@@ -50,13 +53,27 @@ def _serialize_args_and_kwargs_for_eager_mode(
             args, kwargs = serialization.loads(data, content_type, content_encoding)
         return args, kwargs
 
+    def original_apply_async(self, *args, **kwargs):
+        return super().apply_async(*args, **kwargs)
+
     def apply_async(self, args=None, kwargs=None, **options):
         if app.conf.task_always_eager:
             args, kwargs = self._serialize_args_and_kwargs_for_eager_mode(
                 args=args, kwargs=kwargs, **options
             )
-
-        return super().apply_async(args=args, kwargs=kwargs, **options)
+            # In eager mode, immediately call original apply async as we are
+            # using eager mode for tests, where no transaction is ever actually
+            # committed so transaction.on_commit() is never called.
+            return self.original_apply_async(args=args, kwargs=kwargs, **options)
```

Comment:
Not really sure what the impact of this is, but we sometimes return the value of the caller, and sometimes not? Is it not possible to use a lambda or some wrapper function to assign the value of self.original_apply_async  and return that? Or do we have to return in the other case? Why not have consistent return behavior?

Reply:
I agree it's inconsistent, but we can't return anything in the transaction path: `transaction.on_commit()` registers a callback to be executed at a later time, and returns immediately.

The same issue was already present in django-post-request-task: https://github.com/mozilla/django-post-request-task/blob/4f1e33ec778bbba6210f00c7eb6d5820e0bc8d82/post_request_task/task.py#L98-L104 (`_append_task()` returns `None`)

Reply:
Yes but then if we have to return none, why not also return none in the always_eager mode.

```python
self.original_apply_async(args=args, kwargs=kwargs, **options)
return None
```

🤷 just seems like we are creating different behaviour in tests, which is already different. I'm not strongly advocating for this, just don't see why not if it's 1LOC

Reply:
Ah yes, we could probably do that. It was a bit more controversial to do in django-post-request-task, which was somewhat generic (so it could have consumers of the library that wanted the return value in some cases) but we should be OK to do it now.

Reply:
ddd44bade557c76c4cfc9e7887e67d7d58d1f859